## Concepts Covered
- [Standard Error of the Sample Mean](#standard-error-of-the-sample-mean)
- [Z-Score and Standartization](#Z-Score-and-Standartization)
- [Confidence Level and Confidence Interval](#confidence-level-and-confidence-interval)
- [Confidence Intervals for Large Samples](#confidence-intervals-for-large-samples)
- [Degrees of Freedom](#degrees-of-freedom)
- [Confidence Intervals for Small Samples](#confidence-intervals-for-small-samples)

In [1]:
import pandas as pd
import seaborn as sns
import scipy.stats as st
import math
import statsmodels.stats.api as sm
from sklearn import preprocessing
import numpy as np



In [2]:
data_path = "Inf_Stats_files/Confidence Intervals/pizza_restaurant.csv"
df = pd.read_csv(data_path)
df.head()

,Product Name,Crust,Toppings,Price,Delivery Time,# pizzas the customer ordered before
0,Pepperoni,Cheese_and_Garlic,4,17,26.3,4
1,Hawaiian,Cheese_and_Chili,4,17,27.8,4
2,Calzone,Cheese_and_Garlic,3,20,31.5,7
3,Margherita,Cheese,4,23,20.8,7
4,Calzone,Cheese_and_Garlic,4,19,27.7,8


In [3]:
df.describe()

,Toppings,Price,Delivery Time,# pizzas the customer ordered before
count,1000.000000,1000.000000,1000.000000,1000.000000
mean,3.965000,19.342000,25.061100,5.659000
std,1.021185,3.345479,2.490397,2.459831
min,1.000000,12.000000,17.800000,0.000000
25%,3.000000,17.000000,23.300000,4.000000
50%,4.000000,19.000000,25.100000,5.000000
75%,5.000000,21.000000,26.700000,7.000000
max,7.000000,33.000000,32.400000,15.000000


# Standard Error of the Sample Mean
It estimates how far the sample mean is likely to be from the population mean.

Standard Error is calculated as follows:
$$
SE = \frac{\sigma}{\sqrt{n}}
$$
where $\sigma$ is standard deviation of the sample, $n$ is number of observations.

[Back to top](#concepts-covered)


In [4]:
# SE of price
# through formula:
print(df.Price.std() / math.sqrt(df.Price.count())) 
# with stats package:
print(st.sem(df.Price))

0.105793327900337
0.105793327900337


# Z-Score and Standartization

Z-score estimates how far away a data point is from its mean. It represents the number of standard deviations an observation is from the mean.
$$
z = \frac{x - \mu}{\sigma}
$$

[Back to top](#concepts-covered)


In [5]:
# Z-score for delivery time
# using formula
df["Delivery Time standardized"] = (df["Delivery Time"] - df["Delivery Time"].mean())/df["Delivery Time"].std()

In [6]:
# using sklearn
df["Delivery Time standardized2"] = preprocessing.scale(df["Delivery Time"])

In [7]:
# Comapre the results
df.head()

,Product Name,Crust,Toppings,Price,Delivery Time,# pizzas the customer ordered before,Delivery Time standardized,Delivery Time standardized2
0,Pepperoni,Cheese_and_Garlic,4,17,26.3,4,0.497471,0.497720
1,Hawaiian,Cheese_and_Chili,4,17,27.8,4,1.099784,1.100335
2,Calzone,Cheese_and_Garlic,3,20,31.5,7,2.585491,2.586785
3,Margherita,Cheese,4,23,20.8,7,-1.711012,-1.711868
4,Calzone,Cheese_and_Garlic,4,19,27.7,8,1.059630,1.060160


In [8]:
df.describe()

,Toppings,Price,Delivery Time,# pizzas the customer ordered before,Delivery Time standardized,Delivery Time standardized2
count,1000.000000,1000.000000,1000.000000,1000.000000,1.000000e+03,1.000000e+03
mean,3.965000,19.342000,25.061100,5.659000,1.385558e-16,1.421085e-16
std,1.021185,3.345479,2.490397,2.459831,1.000000e+00,1.000500e+00
min,1.000000,12.000000,17.800000,0.000000,-2.915639e+00,-2.917098e+00
25%,3.000000,17.000000,23.300000,4.000000,-7.071562e-01,-7.075101e-01
50%,4.000000,19.000000,25.100000,5.000000,1.562000e-02,1.562781e-02
75%,5.000000,21.000000,26.700000,7.000000,6.580878e-01,6.584171e-01
max,7.000000,33.000000,32.400000,15.000000,2.946879e+00,2.948354e+00


In [9]:
# Values vary due to the difference in standard deviation calculated.
# To obtain the same result as in sklearn.preprocessing.scale(), in the formula # we need to pass ddof=0 to use standard deviation of the population.

df["Delivery Time standardized"] = (df["Delivery Time"] - df["Delivery Time"].mean())/df["Delivery Time"].std(ddof=0)
df.describe()

,Toppings,Price,Delivery Time,# pizzas the customer ordered before,Delivery Time standardized,Delivery Time standardized2
count,1000.000000,1000.000000,1000.000000,1000.000000,1.000000e+03,1.000000e+03
mean,3.965000,19.342000,25.061100,5.659000,1.421085e-16,1.421085e-16
std,1.021185,3.345479,2.490397,2.459831,1.000500e+00,1.000500e+00
min,1.000000,12.000000,17.800000,0.000000,-2.917098e+00,-2.917098e+00
25%,3.000000,17.000000,23.300000,4.000000,-7.075101e-01,-7.075101e-01
50%,4.000000,19.000000,25.100000,5.000000,1.562781e-02,1.562781e-02
75%,5.000000,21.000000,26.700000,7.000000,6.584171e-01,6.584171e-01
max,7.000000,33.000000,32.400000,15.000000,2.948354e+00,2.948354e+00


# Confidence Level and Confidence Interval

<b>Confidence level</b> indicates how sure one can be that a statistical conclusion is correct. Expressed in $\%$.

A typical confidence level is set at 95% chance that the results are accurate.

Confidence level is used in hypothesis testing and is used to calculate confidence intervals.

High confidence level doesn't mean higher accuracy, just higher certainty.

<b>Confidence interval</b> is a range likely to contain the true population parameter with a certain level certainty.

Confidence interval width is determined by:
- Standard error of the sample mean
- Confidence level

[Back to top](#concepts-covered)

In [10]:
cl = [0.005, 0.025, 0.05, 0.95, 0.975, 0.995]
for alpha in cl:
    print(f"The corresponding value in normal distribution for {alpha} is: {round(st.norm.ppf(alpha), 3)}")

The corresponding value in normal distribution for 0.005 is: -2.576
The corresponding value in normal distribution for 0.025 is: -1.96
The corresponding value in normal distribution for 0.05 is: -1.645
The corresponding value in normal distribution for 0.95 is: 1.645
The corresponding value in normal distribution for 0.975 is: 1.96
The corresponding value in normal distribution for 0.995 is: 2.576


In [11]:
# Inverse of ppf:
st.norm.cdf(1.96)

np.float64(0.9750021048517795)

# Confidence Intervals for Large Samples

$$
CI = \bar{x} \pm z \frac{s}{\sqrt{n}}
$$

Sample size must be above 30 (ideally above 100) observations.

Practical advice: not to make conclusions on samples of 10 or less observations.

Z-value represents the desired confidence level (often 95%). Confidence interval: from 2.5 to 97.5 percentile. To determine CI, find z value corresponding to 2.5 or 97.5% instead of 95%

<i>Example.</i>

A survey shows that 60% of customers would buy the new product. The company wants to know the survey's accuracy. Confidence interval can estimate the range of possible true populaton percentages. For instance, 95% confidence interval of 55-65% would mean 95% confidence that the true percentage of customers falls within that range. This helps recogizing the level of uncertainty in the predictions and evaluate the potential risks.

[Back to top](#concepts-covered)

In [12]:
# Confidence interval for the Price mean
print(f"The mean is {df.Price.mean()}")
print(f"The 95-confidence interval for it: {st.norm.interval(confidence=0.95,
                                            loc = df.Price.mean(),
                                            scale = st.sem(df.Price))}")

The mean is 19.342
The 95-confidence interval for it: (np.float64(19.134648887510703), np.float64(19.549351112489294))


In [13]:
# function to compute the confidence interval for the mean
# for each numerical variable that has a sample size bigger than 30
def conf_interval_for_mean(X: pd.Series, confidence=0.95)->tuple[float, float]:
    if not isinstance(X, pd.Series):
        raise TypeError("X must be a pandas Series.")
    if not 0 < confidence < 1:
        raise ValueError("confidence must be between 0 and 1.")
    if X.count() < 30:
        raise ValueError(f"At least 30 non-null observations are required, got {X.count()}.")

    x = X.dropna()
    mean = x.mean()
    se = st.sem(x)

    ci_low, ci_high = st.norm.interval(confidence=confidence, loc=mean, scale=se)
    return (float(ci_low), float(ci_high))

In [14]:
def calculate_confidence_intervals(df, confidence=0.95):
    intervals = {}

    # Iterate over each numerical column in the dataframe
    for column in df.select_dtypes(include=[np.number]).columns:
        data = df[column].dropna()
        # Check if the sample size is greater than 30
        if len(data) > 30:
            # Compute the confidence interval for the mean
            interval = st.norm.interval(confidence, loc=data.mean(), scale=st.sem(data))
            intervals[column] = interval
    return intervals


print(calculate_confidence_intervals(df))

{'Toppings': (np.float64(3.9017074909279676), np.float64(4.028292509072032)), 'Price': (np.float64(19.134648887510703), np.float64(19.549351112489294)), 'Delivery Time': (np.float64(24.9067464105456), np.float64(25.2154535894544)), '# pizzas the customer ordered before': (np.float64(5.5065408812039385), np.float64(5.811459118796061)), 'Delivery Time standardized': (np.float64(-0.06201051624377051), np.float64(0.06201051624377079)), 'Delivery Time standardized2': (np.float64(-0.06201051624377051), np.float64(0.06201051624377079))}


# Degrees of Freedom
The number of uncostrained variables that can vary in a dataset, or number of "free" variables.

DoF is a vital concept for hypothesis testing, statistical modelling, and comprehending distribution behavior.

<i>Example</i>. We know that the average score of 5 students is 80%. That means that for individual scores, 4 students can score anything, while 5th score is constrained by those 4 and the average. Here, we have 4 degrees of freedom.

Confidence intervals help estimate population parameters using sample data. To construct a confidence interval for population mean using t-distribution (the t-distribution is used when working with a small sample rather than the entire population), we use both sample mean and variance (via standard error). Here, $DoF = (n-1)$ where $n$ is the sample size. Why? We estimate two parameters from the sample data - sample mean $\bar{x}$ (point estimate of $\mu$) and sample variance. Sample mean is calculated without any constraints and can take any value. To calculate sample variance, we consider differences between each data point and sample mean - there's constraint on the remaining points. So DoF (n-1) for t-distribution is derived from sample variance as it's the only parameter with constraint.

[Back to top](#concepts-covered)

# Confidence Intervals for Small Samples

Small samples are samples with 30 or fewer observations.

The <b>Margin of Error</b> is the expected maximum deviation the results are expected to differ from the population. The smaller the sample size, the larger the margin of error.

Unlike z-distribution used for large samples, t-distribution is used to calculate confidence intervals for small samples. t-distribution values vary based on the degrees of freedon.

[Back to top](#concepts-covered)

In [18]:
# create small sample from the whole sample dataset
sample = df.sample(20)
sample.describe()

,Toppings,Price,Delivery Time,# pizzas the customer ordered before,Delivery Time standardized,Delivery Time standardized2
count,20.000000,20.000000,20.000000,20.000000,20.000000,20.000000
mean,3.600000,18.950000,24.825000,6.450000,-0.094852,-0.094852
std,1.046297,3.394655,2.110344,2.605157,0.847817,0.847817
min,2.000000,13.000000,21.800000,3.000000,-1.310125,-1.310125
25%,3.000000,16.750000,23.200000,4.000000,-0.747684,-0.747684
50%,3.000000,18.500000,24.900000,6.000000,-0.064721,-0.064721
75%,4.000000,21.250000,25.925000,8.000000,0.347066,0.347066
max,6.000000,25.000000,29.900000,11.000000,1.943996,1.943996


In [19]:
# Confidence interval for the Price mean
print(f"The mean is {sample.Price.mean()}")
print(f"The 95-confidence interval for it: {st.t.interval(confidence=0.95,
                                            df = len(sample)-1,
                                            loc = sample.Price.mean(),
                                            scale = st.sem(sample.Price))}")

The mean is 18.95
The 95-confidence interval for it: (np.float64(17.361252442019314), np.float64(20.538747557980685))


In [20]:
# Confidence intervals for each numerical value with a sample size <= 30
def calculate_confidence_intervals(df, confidence=0.95):
    intervals = {}

    # Iterate over each numerical column in the dataframe
    for column in df.select_dtypes(include=[np.number]).columns:
        data = df[column].dropna()
        if len(data) <= 30:
            # Compute the confidence interval for the mean
            interval = st.t.interval(confidence, df=len(data)-1, 
                                     loc=data.mean(), scale=st.sem(data))
            intervals[column] = interval
    return intervals


print(calculate_confidence_intervals(sample))

{'Toppings': (np.float64(3.1103180581113437), np.float64(4.0896819418886565)), 'Price': (np.float64(17.361252442019314), np.float64(20.538747557980685)), 'Delivery Time': (np.float64(23.837328513103344), np.float64(25.812671486896647)), '# pizzas the customer ordered before': (np.float64(5.230749072903038), np.float64(7.669250927096963)), 'Delivery Time standardized': (np.float64(-0.4916419626852147), np.float64(0.3019387891224176)), 'Delivery Time standardized2': (np.float64(-0.4916419626852147), np.float64(0.3019387891224176))}
